src/graph/on_disk_graph.cpp

```diff
@@ -240,22 +251,31 @@ std::unique_ptr<VertexScanState> OnDiskGraph::prepareVertexScan(TableCatalogEntr
     return std::make_unique<OnDiskGraphVertexScanState>(*context, tableEntry, propertiesToScan);
 }
 
-bool OnDiskGraphNbrScanState::InnerIterator::next(evaluator::ExpressionEvaluator* predicate) {
-    bool hasAtLeastOneSelectedValue = false;
+bool OnDiskGraphNbrScanState::InnerIterator::next(evaluator::ExpressionEvaluator* predicate,
+    semi_mask_t* nbrNodeMask_) {
     do {
         restoreSelVector(*tableScanState->outState);
         if (!relTable->scan(context->getTransaction(), *tableScanState)) {
             return false;
         }
         saveSelVector(*tableScanState->outState);
         if (predicate != nullptr) {
-            hasAtLeastOneSelectedValue =
-                predicate->select(tableScanState->outState->getSelVectorUnsafe(),
-                    !tableScanState->outState->isFlat());
-        } else {
-            hasAtLeastOneSelectedValue = tableScanState->outState->getSelVector().getSelSize() > 0;
+            predicate->select(tableScanState->outState->getSelVectorUnsafe(),
+                !tableScanState->outState->isFlat());
+        }
+        if (nbrNodeMask_ != nullptr) {
+            auto selectedSize = 0u;
+            auto buffer = tableScanState->outState->getSelVectorUnsafe().getMutableBuffer();
+            for (auto i = 0u; i < tableScanState->outState->getSelSize(); ++i) {
+                auto pos = tableScanState->outState->getSelVector()[i];
+                buffer[selectedSize] = pos;
+                auto nbrNodeID = tableScanState->outputVectors[0]->getValue<nodeID_t>(pos);
+                selectedSize += nbrNodeMask_->isMasked(nbrNodeID.offset);
+            }
+            tableScanState->outState->getSelVectorUnsafe().setSelSize(selectedSize);
+            tableScanState->outState->getSelVectorUnsafe().setToFiltered();
```

Comment:
You can simply do `tableScanState->outState->getSelVectorUnsafe().setToFiltered(selectedSize);`